src/qr.jl

```diff
@@ -538,8 +538,10 @@ end
 function rank(A::QRPivoted; atol::Real=0, rtol::Real=min(size(A)...) * eps(real(float(one(eltype(A.Q))))) * iszero(atol))
     m = min(size(A)...)
     m == 0 && return 0
-    tol = max(atol, rtol*abs(A.R[1,1]))
-    return something(findfirst(i -> abs(A.R[i,i]) <= tol, 1:m), m+1) - 1
+    rdiag = diag(getfield(A, :factors))
+    tol = max(atol, rtol*abs(rdiag[1]))
+
+    return something(findfirst(abs.(rdiag) .<= tol), m+1) - 1
```

Comment:
This also makes unnecessary temporary arrays compared to the original.

Reply:
@stevengj the original code makes many more allocations. I document this in the associated issue

Reply:
Each call to A.R actually allocates more memory and slows things down

Reply:
Yes, I understand that `A.R` inadvertently allocates.  But if you replace it with `factors[i,i]` it should no longer allocate, no?

Reply:
correct